pytorch_lightning/callbacks/early_stopping.py

```diff
@@ -0,0 +1,121 @@
+import logging as log
+import warnings
+
+import numpy as np
+
+from .callback import Callback
+
+
+class EarlyStopping(Callback):
+    r"""
+    Stop training when a monitored quantity has stopped improving.
+
+    Args:
+        monitor (str): quantity to be monitored. Default: ``'val_loss'``.
+        min_delta (float): minimum change in the monitored quantity
+            to qualify as an improvement, i.e. an absolute
+            change of less than `min_delta`, will count as no
+            improvement. Default: ``0``.
+        patience (int): number of epochs with no improvement
+            after which training will be stopped. Default: ``0``.
+        verbose (bool): verbosity mode. Default: ``0``.
+        mode (str): one of {auto, min, max}. In `min` mode,
+            training will stop when the quantity
+            monitored has stopped decreasing; in `max`
+            mode it will stop when the quantity
+            monitored has stopped increasing; in `auto`
+            mode, the direction is automatically inferred
+            from the name of the monitored quantity. Default: ``'auto'``.
+        strict (bool): whether to crash the training if `monitor` is
+            not found in the metrics. Default: ``True``.
+
+    Example::
+
+        from pytorch_lightning import Trainer
+        from pytorch_lightning.callbacks import EarlyStopping
+
+        early_stopping = EarlyStopping('val_loss')
+        Trainer(early_stop_callback=early_stopping)
+    """
+
+    def __init__(self, monitor='val_loss',
+                 min_delta=0.0, patience=0, verbose=0, mode='auto', strict=True):
+        super(EarlyStopping, self).__init__()
+
+        self.monitor = monitor
+        self.patience = patience
+        self.verbose = verbose
+        self.strict = strict
+        self.min_delta = min_delta
+        self.wait = 0
+        self.stopped_epoch = 0
+
+        if mode not in ['auto', 'min', 'max']:
```

Comment:
I would work rather with enums as we do here:
 https://github.com/PyTorchLightning/pytorch-lightning/pull/770/files#r380354936

Reply:
I like enum too but for internal use. I think exposing an enum to the user is not a good pattern because it requires yet-another-object to import and use. It's much simpler to use a string. Also, things such as `min` and `max` are quite common (see Keras for example). What do you think?

Reply:
ok and what about OpenCV style, having constants with an int...
```
BEST_MODE_MIN = 1
BEST_MODE_MAX = 2
BEST_MODE_AUTO = 3
```
and later in code, you use something like `pytorch_ligntning.BEST_MODE_MAX`
just feel like strings are too sensitive to typos or miss-abbreviations :]

Reply:
The issue remains the same (but I admit it would be simpler than an enum).

I tend to serialize all my parameters into a yaml/json file when running a job and having those kinds of args would make it harder.

That being said I am only a contributor so it's up to you to decide.

Reply:
Anyone will to help is very welcome! I was just a contributor couple months ago too... That is the great thing about open source =)

Reply:
I think passing strings is fine. We’re already doing it in a number of other places, e.g. ddp backend for trainer.

Reply:
Agree. Then, to be honest, I really think in terms of UX using a string, in that case, is the best pattern as it lowers the barrier usage and makes it easier to serialize.

Reply:
I'll mark as resolved if anyone wants to use enum or enum-like in that kind of situation may be an issue should be open about it.

Reply:
Ok let's keep string for now..
@neggert I do not think that justifications like we already does it elsewhere is good if may do it better... But let's talk about it in separate issue and decide later =)

Reply:
Let's give you a example from my motivation.. You have a huge model, you make typo in string, you launch it before you go home hoping having results other day morning... then it runs for a while and crash because of this typo and in the morning you see just the typo error lol